src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.cs

```diff
@@ -253,23 +264,58 @@ private RequestContext CreateRequestContext(QueueItem queueItem)
             // There are multiple possible solutions that we could be interested in, so we need to find the document
             // first and then get the solution from there. If we're not given a document, this will return the default
             // solution
-            var (documentId, solution) = _solutionProvider.GetDocumentAndSolution(queueItem.TextDocument, queueItem.ClientName);
+            var (documentId, workspaceSolution) = _solutionProvider.GetDocumentAndSolution(queueItem.TextDocument, queueItem.ClientName);
 
-            // Now we can update the solution to represent the LSP view of the world, with any text changes we received
-            solution = GetSolutionWithReplacedDocuments(solution);
+            var lspSolution = GetLSPSolution(workspaceSolution);
 
             // If we got a document id back, we pull it out of our updated solution so the handler is operating on the latest
             // document text. If document id is null here, this will just return null
-            var document = solution.GetDocument(documentId);
+            var document = lspSolution.GetDocument(documentId);
 
-            // Logically, if a mutating request fails we don't want to take its mutations, so giving it the "real"
-            // tracker is a bad idea, but since we tear down the queue for any error anyway, the document tracker
-            // will be emptied and no future requests will be handled, so we don't need to do anything special here.
             var trackerToUse = queueItem.MutatesSolutionState
                 ? (IDocumentChangeTracker)_documentChangeTracker
                 : new NonMutatingDocumentChangeTracker(_documentChangeTracker);
 
-            return new RequestContext(solution, queueItem.ClientCapabilities, queueItem.ClientName, document, trackerToUse);
+            return new RequestContext(lspSolution, queueItem.ClientCapabilities, queueItem.ClientName, document, trackerToUse);
+        }
+
+        /// <summary>
+        /// Gets the "LSP view of the world", either by forking the workspace solution and updating the documents we track
+        /// or by simply returning our cached solution if it is still valid.
+        /// </summary>
+        private Solution GetLSPSolution(Solution workspaceSolution)
+        {
+            var workspace = workspaceSolution.Workspace;
+
+            var recalculateSolution = false;
+            // if we don't have a cached LSP view of the world, we need to create a new one
+            if (!_lastLSPSolution.ContainsKey(workspace))
```

Comment:
this whole method seems very complex.  I would expect it to be more like:

```c#
            // We can reuse an existing lsp solution if:
            // 1. we've already computed it for this workspace.
            // 2. the workspace's solution itself hasn't changed in the meantime
            if (!_lastLSPSolution.TryGetValue(workspace, out var lspSolution) || 
                !_lastWorkspaceSolution.TryGetValue(workspace, out var lastSolution) ||
                lastSolution != workspaceSolution)
            {
                lspSolution = GetSolutionWithReplacedDocuments(workspaceSolution);
                _lastWorkspaceSolution[workspace] = workspaceSolution;
                _lastLSPSolution[workspace] = lspSolution;
            }
```

Reply:
Alternatively, instead of two dictionaries, just have a `Dictionary<Workspace, (Solution lastWorkspaceSolution, Solution lastLspSolution)>`.  Then you can just see the last thing you cached there, and if hte workspace-solution changed.  If not, you use the lastLspSolution.  If so, you recompute it and cache it.

Reply:
Ironically, I did this "long hand" because I was sure you would complain if I tried to cram it into one if statement 😛

I think the single dictionary might make it easier, but it's not quite as simple as that if statement you wrote because we don't want to short circuit and avoid the workspace check, but it might just be a matter of re-ordering. Will take a look. I wasn't worried too much about this, because the dictionaries are only necessary while we don't have server-per-workspace.

Reply:
> Ironically, I did this "long hand" because I was sure you would complain if I tried to cram it into one if statement 

Understood.  In general though, i think it's idiomatic to state:

```
if (! logic-that-says-i-can-reuse-from-cache)
{
    // compute new data.
    // store in cache
}

return value;
```

it's easier for me to understnad that pattern versus having a boolean that states if we need to recopmute that is then mutated through several paths. :)

Reply:
> but it's not quite as simple as that if statement you wrote because we don't want to short circuit and avoid the workspace check

Why?

It's an `||`.  So we only ever reuse when all our conditions pass.  if any conditions fail, we recompute and recache.  Which seems entirely correct and appropriate to me.

Reply:
err.. yeah.. I don't know what I was thinking